### PR TITLE
8294316: SA core file support is broken on macosx-x64 starting with macOS 12.x

### PR DESCRIPTION
**Pull request by tkrodriguez**
This is a minor fix to core file reading on macos x.  I can confirm that after this fix I can run the problem listed SA core file tests on Ventura.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294316](https://bugs.openjdk.org/browse/JDK-8294316): SA core file support is broken on macosx-x64 starting with macOS 12.x (**Bug** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [bd45a311](https://git.openjdk.org/jdk/pull/14569/files/bd45a3114891536d68ee7a6329127193040f8769)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [2ebe4d09](https://git.openjdk.org/jdk/pull/14569/files/2ebe4d0986866e908ead53feebae2ea2d53bd67b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14569/head:pull/14569` \
`$ git checkout pull/14569`

Update a local copy of the PR: \
`$ git checkout pull/14569` \
`$ git pull https://git.openjdk.org/jdk.git pull/14569/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14569`

View PR using the GUI difftool: \
`$ git pr show -t 14569`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14569.diff">https://git.openjdk.org/jdk/pull/14569.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14569#issuecomment-1599276254)

### CODE REVIEW

**Review by plummercj — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

File: src/jdk.hotspot.agent/macosx/native/libsaproc/ps_core.c [hunk 1]
```diff
@@ -297,13 +297,17 @@ static bool read_core_segments(struct ps_prochandle* ph) {
...
+      // The base of the library is offset by a random amount which ends up as a load command with a
+      // filesize of 0.  This must be ignored otherwise the base address of the library is wrong.
+      if (segcmd.filesize != 0) {
+        if (add_map_info(ph, fd, segcmd.fileoff, segcmd.vmaddr, segcmd.vmsize, segcmd.flags) == NULL) {
+          print_debug("Failed to add map_info at i = %d\n", i);
+          goto err;
+        }
+        print_debug("LC_SEGMENT_64 added: nsects=%d fileoff=0x%llx vmaddr=0x%llx vmsize=0x%llx filesize=0x%llx %s\n",
+                    segcmd.nsects, segcmd.fileoff, segcmd.vmaddr, segcmd.vmsize,
+                    segcmd.filesize, &segcmd.segname[0]);
```
[the remark this answers is not in the document]
Comment (tkrodriguez) [reply src/jdk.hotspot.agent/macosx/native/libsaproc/ps_core.c]: Good idea.  I've flattened the ifs a little and fixed the message.

**Review by sspitsyn — APPROVED** [review]
Update looks good.
Thanks,
Serguei

File: src/jdk.hotspot.agent/macosx/native/libsaproc/ps_core.c [hunk 2]
```diff
@@ -297,11 +297,14 @@ static bool read_core_segments(struct ps_prochandle* ph) {
         print_debug("failed to read LC_SEGMENT_64 i = %d!\n", i);
         goto err;
       }
-      if (add_map_info(ph, fd, segcmd.fileoff, segcmd.vmaddr, segcmd.vmsize, segcmd.flags) == NULL) {
+      // The base of the library is offset by a random amount which ends up as a load command with a
+      // filesize of 0.  This must be ignored otherwise the base address of the library is wrong.
+      if (segcmd.filesize != 0 && add_map_info(ph, fd, segcmd.fileoff, segcmd.vmaddr, segcmd.vmsize, segcmd.flags) == NULL) {
```
Comment (plummercj) [inline src/jdk.hotspot.agent/macosx/native/libsaproc/ps_core.c:302]: I actually preferred this part with  two `if` statements. Seems easier to read that way.

Comment (tkrodriguez) [reply src/jdk.hotspot.agent/macosx/native/libsaproc/ps_core.c:302]: Ok.  I reverted it to 2 ifs.

**Review by plummercj — APPROVED** [review]

**Review by plummercj — APPROVED** [review]
Please update the copyright before integrating.

**Comment by tkrodriguez** [comment]
Thanks!

/integrate

**Pull request closed without merging**